File: minesweeper/java-minesweeper/src/test/java/it/chicio/minesweeper/field/parser/FieldRowContentHeaderChainParserTest.java
package it.chicio.minesweeper.field.parser;

import it.chicio.minesweeper.field.Field;
import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.not;
import static org.junit.Assert.assertThat;

public class FieldRowContentHeaderChainParserTest {
    private FieldRowContentHeaderChainParser fieldRowContentHeaderChainParser;

    @Before
    public void setUp() {
        fieldRowContentHeaderChainParser = new FieldRowContentHeaderChainParser();
    }

    @Test
    public void canParseValidHeaderRow() {
        assertThat(fieldRowContentHeaderChainParser.canParse("1 1"), is(true));
    }

    @Test
    public void canNotParseValidHeaderRow() {
        assertThat(fieldRowContentHeaderChainParser.canParse("1"), is(false));
        assertThat(fieldRowContentHeaderChainParser.canParse(""), is(false));
    }

    @Test
    public void parseHeader() {
        Field previousField = new Field(new String[][]{{"*"}});

        FieldsParsingStatus newFieldsParsingStatus = fieldRowContentHeaderChainParser.parse(
                new FieldsParsingStatusBuilder()
                        .withCurrentRowContent("2 2")
                        .withHeaderNumberOfRowsForCurrentField(1)
                        .withFieldsParsed(new ArrayList<Field>())
                        .withCurrentField(previousField)
                        .build()
        );

        assertThat(newFieldsParsingStatus.fieldsParsed.size(), is(1));
        assertThat(newFieldsParsingStatus.currentField.numberOfColumn(), is(2));
        assertThat(newFieldsParsingStatus.currentField.numberOfRows(), is(2));
        assertThat(newFieldsParsingStatus.currentField, is(not(previousField)));
    }

    @Test(expected = RuntimeException.class)
    public void failParseForInvalidField() {
        Field previousField = new Field(new String[][]{});

        fieldRowContentHeaderChainParser.parse(
                new FieldsParsingStatusBuilder()
                        .withCurrentRowContent("2 2")
                        .withCurrentField(new Field(new String[][]{{"*"}}))
                        .withHeaderNumberOfRowsForCurrentField(1)
                        .withFieldsParsed(new ArrayList<Field>())
                        .withCurrentField(previousField)
                        .build()
        );
    }
}

File: minesweeper/kotlin-mockk-minesweeper/src/test/kotlin/it/chicio/minesweeper/field/parser/FieldRowContentHeaderChainParserTest.kt
package it.chicio.minesweeper.field.parser

import it.chicio.minesweeper.FieldFactory
import it.chicio.minesweeper.FieldsParsingStatusBuilder
import it.chicio.minesweeper.field.Field
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.util.*

@DisplayName("FieldRowContentHeaderChainParser")
class FieldRowContentHeaderChainParserTest {
    private lateinit var fieldRowContentHeaderChainParser: FieldRowContentHeaderChainParser

    @BeforeEach
    fun setUp() {
        fieldRowContentHeaderChainParser = FieldRowContentHeaderChainParser()
    }

    @Test
    fun `can parse valid header row`() =
            assertTrue(fieldRowContentHeaderChainParser.canParse("1 1"))

    @Test
    fun `can not parse valid header row`() {
        assertFalse(fieldRowContentHeaderChainParser.canParse("1"))
        assertFalse(fieldRowContentHeaderChainParser.canParse(""))
    }

    @Test
    fun `parse header`() {
        val previousField = FieldFactory().make(arrayOf(arrayOf("*")))
        val newFieldsParsingStatus = fieldRowContentHeaderChainParser.parse(
                FieldsParsingStatusBuilder(
                        currentField = previousField,
                        currentRowContent = "2 2",
                        headerNumberOfRowsForCurrentField = 1,
                ).build()
        )
        assertEquals(newFieldsParsingStatus.fieldsParsed.size, 1)
        assertEquals(newFieldsParsingStatus.currentField!!.numberOfColumn, 2)
        assertEquals(newFieldsParsingStatus.currentField!!.numberOfRows, 2)
        assertNotEquals(newFieldsParsingStatus.currentField, previousField)
    }

    @Test
    fun `fail parse for invalid field`() {
        assertThrows(RuntimeException::class.java) {
            fieldRowContentHeaderChainParser.parse(
                    FieldsParsingStatusBuilder(
                            currentRowContent = "2 2",
                            currentField = FieldFactory().make(arrayOf()),
                            headerNumberOfRowsForCurrentField = 1,
                    ).build()
            )
        }
    }
}